### ISSUE TITLE
Settings allowJs=true in tsconfigs will make debugger step in the javascript files instead of the typescript files.

### ISSUE THREAD

**Issue by SuperCorks** [microsoft/vscode#101927]
Issue Type: <b>Bug</b>

1. Setup a typescript project
2. Set `allowJS = true` in the tsconfigs
3. Setup launch.json to debug typescript
4. Put a breakpoint in a typescript file
5. Run

The debugger will stop on the breakpoint, but not in the TypeScript file, it's gonna stop in the corresponding JavaScript file generated by the `tsc` command.

Please see this stackoverflow question for more details.
https://stackoverflow.com/questions/62763392/debugging-typescript-in-vs-code-steps-into-javascript-instead

VS Code version: Code 1.46.1 (cd9ea6488829f560dc949a8b2fb789f3cdc05f5d, 2020-06-17T21:13:20.174Z)
OS version: Windows_NT x64 10.0.18363

**Comment by SuperCorks** [comment]
This behaviour does not occur in Webstorm.

**Comment by connor4312** [comment]
I'm not able to repro this, can you collect a log file?

/jsDebugLogs

**Comment by SuperCorks** [comment]
It's been a while, let me see if I can reproduce it.

**Comment by SuperCorks** [comment]
Ok... got it to reproduce quite easily actually. 
![image](https://user-images.githubusercontent.com/6114650/98289447-37574400-1f76-11eb-883b-9e57583fab0d.png)

Here's the contents of the log file and a sample project in a zip file that makes the bug happen on my machine.
To run the project, first run `npm install`, then the `tsc` and finally, use the `Launch Program` config to debug is VS Code.

[debugadapter-legacy.txt](https://github.com/microsoft/vscode/files/5496672/debugadapter-legacy.txt)
[vs-code-breakpoint-bug.zip](https://github.com/microsoft/vscode/files/5496656/vs-code-breakpoint-bug.zip)

Version info:

```
Version: 1.46.1 (user setup)
Commit: cd9ea6488829f560dc949a8b2fb789f3cdc05f5d
Date: 2020-06-17T21:13:20.174Z
Electron: 7.3.1
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Windows_NT x64 10.0.19041
```

**Comment by connor4312** [comment]
Thanks for the followup! I noticed you're on an old version of VS Code; can you update to VS Code 1.47 or above? (Preferably to the latest 1.50). In 1.47 we switched to an entirely new debugger which may no longer have this issue.

**Comment by SuperCorks** [comment]
Yes VS Code just updated itself. The is still happening in `v1.50.1`.
Do you want me to regenerate the logs from that version?
```
Version: 1.50.1 (user setup)
Commit: d2e414d9e4239a252d1ab117bd7067f125afd80a
Date: 2020-10-13T15:06:15.712Z
Electron: 9.2.1
Chrome: 83.0.4103.122
Node.js: 12.14.1
V8: 8.3.110.13-electron.0
OS: Windows_NT x64 10.0.19041
```

**Comment by connor4312** [comment]
Please do 🙂

**Comment by connor4312** [comment]
I see what's happening. A small handful people have run into this before. This is because of a behavior that, if you set your `program` to a non-.js file and we find valid sourcemaps referencing it in the outfiles, we'll actually execute the compiled version. This behavior is present because beginners often try to run their .ts files, and it's easy for us to fix that for them.

So we fix this, but keep the args. So although we're running `dist/app.js`, we still run it through ts-node. ts-node replaces the sourcemap in `dist/app.js` with a runtime/ephemeral sourcemap that maps _its_ re-compiled version of app.js back to itself, replacing the correct sourcemap.

tl;dr you can do either of two things:

- Set `outFiles` to an empty array so that sourcemaps are not discovered
- Either drop ts-node and debug compiled files, or use ts-node without manually compiling them
